src/raftstore/store/peer.rs

```diff
@@ -1283,6 +1374,172 @@ impl Peer {
         })))
     }
 
+    fn exec_merge(&mut self,
+                  _ctx: &ExecContext,
+                  req: &AdminRequest)
+                  -> Result<(AdminResponse, Option<ExecResult>)> {
+        PEER_ADMIN_CMD_COUNTER_VEC.with_label_values(&["merge", "all"]).inc();
+
+        let merge_req = req.get_merge();
+        info!("{} merge region {:?} into region {:?}",
+              self.tag,
+              merge_req.get_from_region(),
+              self.region());
+
+        let mut resp = AdminResponse::new();
+        resp.set_merge(MergeResponse::new());
+
+        // update peer meta info
+        let state_key = keys::region_state_key(self.region().get_id());
+        let mut region_local_state = try!(self.engine.get_msg::<RegionLocalState>(&state_key))
+            .unwrap();
+        let mut merge_state = RegionMergeState::new();
+        merge_state.set_state(MergeState::Merging);
+        merge_state.set_from_region(merge_req.get_from_region().clone());
+        region_local_state.set_merge_state(merge_state);
+        try!(self.engine.put_msg(&state_key, &region_local_state));
+        self.merge_state = MergeState::Merging;
+
+        self.start_merging_time = Some(Instant::now());
+
+        PEER_ADMIN_CMD_COUNTER_VEC.with_label_values(&["merge", "success"]).inc();
+
+        Ok((resp,
+            Some(ExecResult::MergeRegion {
+            from_region: merge_req.get_from_region().clone(),
+            into_region: self.region().clone(),
+        })))
+    }
+
+    fn exec_suspend_region(&mut self,
+                           _ctx: &ExecContext,
+                           req: &AdminRequest)
+                           -> Result<(AdminResponse, Option<ExecResult>)> {
+        PEER_ADMIN_CMD_COUNTER_VEC.with_label_values(&["suspend", "all"]).inc();
+
+        let suspend_region = req.get_suspend_region();
+
+        info!("{} suspend region {:?}",
+              self.tag,
+              suspend_region.get_from_region());
+
+        let resp = AdminResponse::new();
+
+        {
+            let from_epoch = suspend_region.get_from_region().get_region_epoch();
+            let latest_epoch = self.region().get_region_epoch();
+            if from_epoch.get_version() < latest_epoch.get_version() ||
+               from_epoch.get_conf_ver() < latest_epoch.get_conf_ver() {
+                return Ok((resp,
+                           Some(ExecResult::RollbackMerge {
+                    from_region: suspend_region.get_from_region().clone(),
+                    into_region: suspend_region.get_into_region().clone(),
+                })));
+            }
+        }
+
+        if self.merge_state != MergeState::Suspended {
```

Comment:
If the `ExecResult::SuspendRegion` is lost due to a node crash at the precise moment, it's possible that this region's state is `MergeState::Suspended` and the same suspend region command will be retried later by the control region. In this case, skip the engine writing since the state is already `MergeState:Suspended`.